progman.1
.\" progman - Copyright 1998-2007 Decklin Foster <decklin@red-bean.com>.
.\" This program is free software; please see LICENSE for details.
.TH "PROGMAN" 1x
.SH "NAME"
progman \- A Window Manager
.SH "SYNOPSIS"
.B progman
.I [options...]
.SH "DESCRIPTION"
.B progman
is an X11 window manager with very few features. It draws a small
titlebar above top-level windows, allows you to manipulate them with the
mouse, and starts other programs when the root window is clicked. That's
also about all it does, so don't expect anything nifty.
.PP
The titlebar contains a small box in the upper right corner. Clicking
the 1st, 2nd, or 3rd buttons on the main area of the titlebar will
raise, move, or lower the window, respectively. Clicking on the box will
respectively hide, resize, or close the window.
.PP
By default, clicking on the root window with button 1 will run
.BR aemenu (1x),
and button 3 will start a new
.BR xterm (1x),
and and button 2 will run
.BR "aemenu \-\-switch" .
.B aemenu
(or another equivalent program) is required to unhide hidden windows;
.B progman
does not do this itself.
.SH "OPTIONS"
.TP
.BI \-\-config \ file\fP, \ \-rc \ file
Read configuration options from
.IR file .
This overrides options which precede it in the command line.
.TP
.BI \-\-font \ xfld\fP, \ \-fn \ xfld
Draw window titles using the font defined by
.IR xfld .
.TP
.BI \-\-bgcolor \ color\fP, \ \-bg \ color
Set the background color of window frames.
.TP
.BI \-\-fgcolor \ color\fP, \ \-fg \ color
Set the foreground color of window frames.
.TP
.BI \-\-bdcolor \ color\fP, \ \-bd \ color
Set the color of window borders.
.TP
.BI \-\-bdwidth \ width\fP, \ \-bw \ width
Draw window borders
.I width
pixels wide.
.TP
.BI \-\-padding \ width\fP, \ \-p \ width
Draw
.I width
pixels of space between window borders and window titles.
.TP
.BI \-\-new1 \ command\fP, \ \-1 \ cmd
Run
.I command
when button 1 is clicked on the root window.
.TP
.BI \-\-new2 \ command\fP, \ \-2 \ cmd
Run
.I command
when button 2 is clicked on the root window.
.TP
.BI \-\-new3 \ command\fP, \ \-3 \ cmd
Run
.I command
when button 3 is clicked on the root window.
.TP
.B \-\-help, \-h
Print a short help message to stdout and exit.
.TP
.B \-\-version, \-v
Print version information to stdout and exit.
.SH "ENVIRONMENT"
.B DISPLAY
Sets which X display will be managed by
.BR progman .
.SH "SEE ALSO"
.BR aeclients (1x),
.BR X (7),
.I The Inter-Client Communication Conventions Manual
(from the X11R6 documentation).
.SH "AUTHOR"
Decklin Foster <decklin@red\-bean.com>


NEWS
1.3.12:

 - Fix wrapping backwards from desk 0.

 - Swap default Button1/Button3 aemenu bindings.

 - Restore the old look of draw_outline (from scratch), undoing 1.3.11's
   changes. Improve and standardize font spacing.

 - Fatten up transient frames a little, in case padding is 0.

1.3.11:

 - Fix issues with atom handling on 64-bit systems.

 - Correct gravity interpretation when we have a large border width.

 - Remove weird resizing algorithm; it was only supposed to be a placeholder
   for something even weirder but I never had time to get the math right. Do
   something more like twm instead (warping the pointer is still evil).

 - When doing any sort of sweep, draw the geometry in the middle of the frame
   and don't draw a line for the title.

 - Save original window border width and restore on exit.

1.3.10:

 - Respect incremental-resize hints when zooming clients, and don't make GTK+
   clients (if they persist their own geom) sometimes forget their zoom state.

 - Removed the shift-click bindings; buttons 4 and 5 now shade/unshade
   and zoom/unzoom in the frame and corner box respectively.

 - Added the ability to map buttons 4 and 5 on the root window. By
   default, they run aedesk to cycle through the available desktops.

 - Added -fa (if applicable) to the usage message.

 - Simplify the initial-position stuff some more, and skip (if enabled)
   interactive mapping in several more cases.

 - Make sure the borders of fullscreen windows are offscreen.

 - Reorganized the source a whole lot.

1.3.9:

 - Redid the initial-position algorithm again.

 - Added support for fullscreen hints.

 - Renamed the mapping options.

 - Fixed drawing of frame outlines when sweeping to the top of the
   screen.

1.3.8:

 - Change the CARD32s to unsigned longs; it seems that Xlib is going to
   handle things correctly. While doing this, rewrote all of the other
   code that deals with atoms and allocating memory for them.

 - Draw the border correctly when moving or resizing a shaded window,
   and don't unshade it as a side effect of other operations.

 - Fix response to EWMH close messages.

 - Reorganized the shared code (and some other stuff), which should
   simplify/speed up the build a bit.

 - Fixed some occasional X errors encountered when switching desktops,
   and a bug where an existing client's desktop could be set incorrectly
   if it was exposed in a certain order.

1.3.7:

 - Added some additional hints for window lists/pagers, and taking
   activation requests.

 - When mapping interactively, clicking with button 3 will map the
   window lowered.

 - Removed support for MWM hints entirely.

 - Fixed aepanel's initial window mapping.

1.3.6:

 - Eliminated the secondary frame box, and UI for setting the desktop
   hint to "all desks" or not. Got rid of the graphical indicators as
   well, since shaded or zoomed state can easily be ascertained by just
   looking at the window. Instead, added a shift-click handler.

 - Changed the operation of resize so that it scales from the center and
   never warps the pointer (which we are not supposed to do anyway).

 - Refactored move and resize so that they don't repeat code. This
   allows us to have interactive mapping essentially for free, so add an
   option for that. And having interactive mapping provides an
   opportunity to rewrite the initial-map geometry code yet again; it
   should be much saner now, although less forgiving of clients which
   try to kludge around the hints instead of doing the right thing.

 - Don't make aedesk -n bail out if there is no number of desktops atom
   set.

 - Mostly rewrote window gravity and resize hint handlers. These should
   do a better job of dealing with most cases.

1.3.5:

 - Allow transient windows to map over struts.

 - Remove the ndesks option and read _NET_NUMBER_OF_DESKTOPS instead.
   Added a -n option to switch-desk to set it, and removed the old -n
   (and -p) in favor of +1 and -1 (or any other integer).

 - Renamed switch-desk to aedesk (since we broke the options anyway).

 - Rewrote README.

1.3.4:

 - Brown paper bag: add missing Xmd.h and limits.h includes.

1.3.3:

 - Use CARD32 everywhere for all atom values. I believe this is what I
   should have done to fix the signed/unsigned confusion. Should work on
   systems where long is 64 bits now.

 - If the current desktop is unset on startup, set it to 0.

 - Don't adjust a PPosition hint relative to any top/left struts, unless
   it's overlapping them. Pretty much every client out there will ignore
   struts and save their position relative to the screen.

 - Get rid of SPACING in aepanel-gtk (you know, Fitts' Law). Also remove
   the frame and the quit button (generally useless). The button for the
   launch menu is now labeled "Launch" for consistency.

 - Make switch-desk -p wrap around properly from desk 0.

 - Added the lib stuff to X11OBJS in clients/Makefile (thanks Doug
   Vetter).

1.3.2:

 - Once again, ignore PPosition hints of 0,0. I don't know how this got
   lost. Perhaps we were doing it wrong all along.

 - If a client has a private colormap, attempt to allocate our colors in
   it before mapping. This doesn't mean the client can't later overwrite
   those colors, but it will at least give us the correct bg intially in
   most cases.

 - When iconifying or de-iconifying a client, also iconify or de-iconify
   all of its transients.

 - Also fix Adam's email in aeclients.1x and NEWS.

 - Add --prev and --next options to switch-desk. This necessitated
   splitting stringatom.c out from atom.c.

 - In aeclient, remove a client from the list when it moves itself to
   another desktop.

1.3.1:

 - Rename this file to NEWS and reformat it.

 - UTF-8 support. This is only, what, 3 years late or so. Requires
   X_HAVE_UTF8_STRING (X.Org or XFree86 >= 4.0.2).

 - Call setlocale on startup (this is necessary for using UTF-8 with
   core fonts or Xaw; it is not strictly so for Xft or GTK+, but

progman.c
/*
 * Copyright 1998-2007 Decklin Foster <decklin@red-bean.com>.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to
 * deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
 * AUTHOR BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
 * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <signal.h>
#include <locale.h>
#include <errno.h>
#include <fcntl.h>
#include <sys/wait.h>
#include <X11/Xatom.h>
#include <X11/cursorfont.h>
#include <X11/extensions/shape.h>
#include "progman.h"
#include "atom.h"
#include "parser.h"

#ifdef HIDPI
#include "close_hidpi.xpm"
#include "shade_hidpi.xpm"
#include "zoom_hidpi.xpm"
#include "unzoom_hidpi.xpm"
#else
#include "close.xpm"
#include "shade.xpm"
#include "zoom.xpm"
#include "unzoom.xpm"
#endif

client_t *head, *focused;
int screen;
unsigned long ndesks = 1;
unsigned long cur_desk = 0;
unsigned int focus_order = 0;
Bool shape;
int shape_event;

XftFont *xftfont;
XftColor xft_fg;
XftColor xft_fg_unfocused;

Colormap def_cmap;
XColor fg;
XColor bg;
XColor fg_unfocused;
XColor bg_unfocused;
XColor button_bg;
XColor bevel_dark;
XColor bevel_light;
XColor bd;
GC pixmap_gc;
GC invert_gc;
Pixmap close_pm;
Pixmap close_pm_mask;
XpmAttributes close_pm_attrs;
Pixmap shade_pm;
Pixmap shade_pm_mask;
XpmAttributes shade_pm_attrs;
Pixmap zoom_pm;
Pixmap zoom_pm_mask;
XpmAttributes zoom_pm_attrs;
Pixmap unzoom_pm;
Pixmap unzoom_pm_mask;
XpmAttributes unzoom_pm_attrs;
Cursor map_curs;
Cursor move_curs;
Cursor resize_n_curs;
Cursor resize_s_curs;
Cursor resize_e_curs;
Cursor resize_w_curs;
Cursor resize_nw_curs;
Cursor resize_sw_curs;
Cursor resize_ne_curs;
Cursor resize_se_curs;

int exitmsg[2];

char *opt_xftfont = DEF_XFTFONT;
char *opt_fg = DEF_FG;
char *opt_bg = DEF_BG;
char *opt_fg_unfocused = DEF_FG_UNFOCUSED;
char *opt_bg_unfocused = DEF_BG_UNFOCUSED;
char *opt_button_bg = DEF_BUTTON_BG;
char *opt_bevel_dark = DEF_BEVEL_DARK;
char *opt_bevel_light = DEF_BEVEL_LIGHT;
char *opt_bd = DEF_BD;
int opt_bw = DEF_BW;
int opt_pad = DEF_PAD;
int opt_bevel = DEF_BEVEL;
char *opt_new[] = { DEF_NEW1, DEF_NEW2, DEF_NEW3, DEF_NEW4, DEF_NEW5 };
int opt_edge_resist = DEF_EDGE_RES;

static void cleanup(void);
static void read_config(char *);
static void setup_display(void);

extern char *__progname;

int
main(int argc, char **argv)
{
	struct sigaction act;
	int ch;

	setlocale(LC_ALL, "");
	read_config(NULL);

	while ((ch = getopt(argc, argv, "c:")) != -1) {
		switch (ch) {
		case 'c':
			read_config(optarg);
			break;
		default:
			printf("usage: %s [-c <config file>]\n", __progname);
			exit(1);
		}
	}

	pipe2(exitmsg, O_CLOEXEC);

	act.sa_handler = sig_handler;
	act.sa_flags = 0;
	sigaction(SIGTERM, &act, NULL);
	sigaction(SIGINT, &act, NULL);
	sigaction(SIGHUP, &act, NULL);
	sigaction(SIGCHLD, &act, NULL);

	setup_display();
	event_loop();
	cleanup();

	return 0;
}

static void
read_config(char *rcfile)
{
	FILE *rc;
	char buf[BUF_SIZE], token[BUF_SIZE], *p;

	if (!(rc = open_rc(rcfile, "progmanrc"))) {
		if (rcfile)
			fprintf(stderr, "%s: rc file '%s' not found\n",
			    __progname, rcfile);
		return;
	}

	while (get_rc_line(buf, sizeof buf, rc)) {
		p = buf;
		while (get_token(&p, token)) {
			if (strcmp(token, "xftfont") == 0) {
				if (get_token(&p, token))
					opt_xftfont = strdup(token);
			} else if (strcmp(token, "fgcolor") == 0) {
				if (get_token(&p, token))
					opt_fg = strdup(token);
			} else if (strcmp(token, "bgcolor") == 0) {
				if (get_token(&p, token))
					opt_bg = strdup(token);
			} else if (strcmp(token, "fgcolor_unfocused") == 0) {
				if (get_token(&p, token))
					opt_fg_unfocused = strdup(token);
			} else if (strcmp(token, "bgcolor_unfocused") == 0) {
				if (get_token(&p, token))
					opt_bg_unfocused = strdup(token);
			} else if (strcmp(token, "button_bgcolor") == 0) {
				if (get_token(&p, token))
					opt_button_bg = strdup(token);
			} else if (strcmp(token, "bdcolor") == 0) {
				if (get_token(&p, token))
					opt_bd = strdup(token);
			} else if (strcmp(token, "bdwidth") == 0) {
				if (get_token(&p, token))
					opt_bw = atoi(token);
			} else if (strcmp(token, "padding") == 0) {
				if (get_token(&p, token))
					opt_pad = atoi(token);
			} else if (strcmp(token, "button1") == 0) {
				if (get_token(&p, token))
					opt_new[0] = strdup(token);
			} else if (strcmp(token, "button2") == 0) {
				if (get_token(&p, token))
					opt_new[1] = strdup(token);
			} else if (strcmp(token, "button3") == 0) {
				if (get_token(&p, token))
					opt_new[2] = strdup(token);
			} else if (strcmp(token, "button4") == 0) {
				if (get_token(&p, token))
					opt_new[3] = strdup(token);
			} else if (strcmp(token, "button5") == 0) {
				if (get_token(&p, token))
					opt_new[4] = strdup(token);
			} else if (strcmp(token, "edgeresist") == 0) {
				if (get_token(&p, token))
					opt_edge_resist = atoi(token);
			}
		}
	}
	fclose(rc);
}

static void
setup_display(void)
{
	XGCValues gv;
	XColor exact;
	XSetWindowAttributes sattr;
	XWindowAttributes attr;
	int shape_err;
	Window qroot, qparent, *wins;
	unsigned int nwins, i;
	client_t *c;

	dpy = XOpenDisplay(NULL);

	if (!dpy) {
		fprintf(stderr, "%s: can't open $DISPLAY \"%s\"\n", __progname,
		    getenv("DISPLAY"));
		exit(1);
	}

	XSetErrorHandler(handle_xerror);
	screen = DefaultScreen(dpy);
	root = RootWindow(dpy, screen);

	map_curs = XCreateFontCursor(dpy, XC_dotbox);
	move_curs = XCreateFontCursor(dpy, XC_fleur);
	resize_n_curs = XCreateFontCursor(dpy, XC_top_side);
	resize_s_curs = XCreateFontCursor(dpy, XC_bottom_side);
	resize_e_curs = XCreateFontCursor(dpy, XC_right_side);
	resize_w_curs = XCreateFontCursor(dpy, XC_left_side);
	resize_nw_curs = XCreateFontCursor(dpy, XC_top_left_corner);
	resize_sw_curs = XCreateFontCursor(dpy, XC_bottom_left_corner);
	resize_ne_curs = XCreateFontCursor(dpy, XC_top_right_corner);
	resize_se_curs = XCreateFontCursor(dpy, XC_bottom_right_corner);

	def_cmap = DefaultColormap(dpy, screen);
	XAllocNamedColor(dpy, def_cmap, opt_fg, &fg, &exact);
	XAllocNamedColor(dpy, def_cmap, opt_bg, &bg, &exact);
	XAllocNamedColor(dpy, def_cmap, opt_fg_unfocused, &fg_unfocused,
	    &exact);
	XAllocNamedColor(dpy, def_cmap, opt_bg_unfocused, &bg_unfocused,
	    &exact);
	XAllocNamedColor(dpy, def_cmap, opt_button_bg, &button_bg, &exact);
	XAllocNamedColor(dpy, def_cmap, opt_bevel_dark, &bevel_dark, &exact);
	XAllocNamedColor(dpy, def_cmap, opt_bevel_light, &bevel_light, &exact);
	XAllocNamedColor(dpy, def_cmap, opt_bd, &bd, &exact);

	XSetLineAttributes(dpy, DefaultGC(dpy, screen), 1, LineSolid, CapButt,
	    JoinBevel);
	XSetFillStyle(dpy, DefaultGC(dpy, screen), FillSolid);

	xft_fg.color.red = fg.red;
	xft_fg.color.green = fg.green;
	xft_fg.color.blue = fg.blue;
	xft_fg.color.alpha = 0xffff;
	xft_fg.pixel = fg.pixel;

	xft_fg_unfocused.color.red = fg_unfocused.red;
	xft_fg_unfocused.color.green = fg_unfocused.green;
	xft_fg_unfocused.color.blue = fg_unfocused.blue;
	xft_fg_unfocused.color.alpha = 0xffff;
	xft_fg_unfocused.pixel = fg_unfocused.pixel;

	xftfont = XftFontOpenName(dpy, DefaultScreen(dpy), opt_xftfont);
	if (!xftfont) {
		fprintf(stderr, "%s: Xft font \"%s\" not found\n", __progname,
		    opt_xftfont);
		exit(1);
	}

	pixmap_gc = XCreateGC(dpy, root, 0, &gv);

	gv.function = GXinvert;
	gv.subwindow_mode = IncludeInferiors;
	invert_gc = XCreateGC(dpy, root,
	    GCFunction | GCSubwindowMode | GCLineWidth, &gv);

	XpmCreatePixmapFromData(dpy, root, close_xpm, &close_pm, &close_pm_mask,
	    &close_pm_attrs);
	XpmCreatePixmapFromData(dpy, root, shade_xpm, &shade_pm, &shade_pm_mask,
	    &shade_pm_attrs);
	XpmCreatePixmapFromData(dpy, root, zoom_xpm, &zoom_pm, &zoom_pm_mask,
	    &zoom_pm_attrs);
	XpmCreatePixmapFromData(dpy, root, unzoom_xpm, &unzoom_pm,
	    &unzoom_pm_mask, &unzoom_pm_attrs);

	utf8_string = XInternAtom(dpy, "UTF8_STRING", False);
	wm_protos = XInternAtom(dpy, "WM_PROTOCOLS", False);
	wm_delete = XInternAtom(dpy, "WM_DELETE_WINDOW", False);
	wm_state = XInternAtom(dpy, "WM_STATE", False);
	wm_change_state = XInternAtom(dpy, "WM_CHANGE_STATE", False);
	net_supported = XInternAtom(dpy, "_NET_SUPPORTED", False);
	net_cur_desk = XInternAtom(dpy, "_NET_CURRENT_DESKTOP", False);
	net_num_desks = XInternAtom(dpy, "_NET_NUMBER_OF_DESKTOPS", False);
	net_client_list = XInternAtom(dpy, "_NET_CLIENT_LIST", False);
	net_client_stack = XInternAtom(dpy, "_NET_CLIENT_LIST_STACKING", False);
	net_active_window = XInternAtom(dpy, "_NET_ACTIVE_WINDOW", False);
	net_close_window = XInternAtom(dpy, "_NET_CLOSE_WINDOW", False);
	net_wm_name = XInternAtom(dpy, "_NET_WM_NAME", False);
	net_wm_desk = XInternAtom(dpy, "_NET_WM_DESKTOP", False);
	net_wm_state = XInternAtom(dpy, "_NET_WM_STATE", False);
	net_wm_state_shaded = XInternAtom(dpy, "_NET_WM_STATE_SHADED", False);
	net_wm_state_mv = XInternAtom(dpy, "_NET_WM_STATE_MAXIMIZED_VERT",
	    False);
	net_wm_state_mh = XInternAtom(dpy, "_NET_WM_STATE_MAXIMIZED_HORZ",
	    False);
	net_wm_state_fs = XInternAtom(dpy, "_NET_WM_STATE_FULLSCREEN", False);
	net_wm_state_rm = 0;
	net_wm_state_add = 1;
	net_wm_state_toggle = 2;
	net_wm_strut = XInternAtom(dpy, "_NET_WM_STRUT", False);
	net_wm_strut_partial = XInternAtom(dpy, "_NET_WM_STRUT_PARTIAL", False);
	net_wm_wintype = XInternAtom(dpy, "_NET_WM_WINDOW_TYPE", False);
	net_wm_type_dock = XInternAtom(dpy, "_NET_WM_WINDOW_TYPE_DOCK", False);
	net_wm_type_menu = XInternAtom(dpy, "_NET_WM_WINDOW_TYPE_MENU", False);
	net_wm_type_splash = XInternAtom(dpy, "_NET_WM_WINDOW_TYPE_SPLASH",
	    False);
	net_wm_type_desk = XInternAtom(dpy, "_NET_WM_WINDOW_TYPE_DESKTOP",
	    False);

	/* The bit about _NET_CLIENT_LIST_STACKING here is an evil lie. */
	append_atoms(root, net_supported, XA_ATOM, &net_cur_desk, 1);
	append_atoms(root, net_supported, XA_ATOM, &net_num_desks, 1);
	append_atoms(root, net_supported, XA_ATOM, &net_client_list, 1);
	append_atoms(root, net_supported, XA_ATOM, &net_client_stack, 1);
	append_atoms(root, net_supported, XA_ATOM, &net_active_window, 1);
	append_atoms(root, net_supported, XA_ATOM, &net_close_window, 1);
	append_atoms(root, net_supported, XA_ATOM, &net_wm_name, 1);
	append_atoms(root, net_supported, XA_ATOM, &net_wm_desk, 1);
	append_atoms(root, net_supported, XA_ATOM, &net_wm_state, 1);
	append_atoms(root, net_supported, XA_ATOM, &net_wm_state_shaded, 1);
	append_atoms(root, net_supported, XA_ATOM, &net_wm_state_mv, 1);
	append_atoms(root, net_supported, XA_ATOM, &net_wm_state_mh, 1);
	append_atoms(root, net_supported, XA_ATOM, &net_wm_state_fs, 1);
	append_atoms(root, net_supported, XA_ATOM, &net_wm_strut, 1);
	append_atoms(root, net_supported, XA_ATOM, &net_wm_strut_partial, 1);
	append_atoms(root, net_supported, XA_ATOM, &net_wm_wintype, 1);
	append_atoms(root, net_supported, XA_ATOM, &net_wm_type_dock, 1);
	append_atoms(root, net_supported, XA_ATOM, &net_wm_type_menu, 1);
	append_atoms(root, net_supported, XA_ATOM, &net_wm_type_splash, 1);
	append_atoms(root, net_supported, XA_ATOM, &net_wm_type_desk, 1);

	get_atoms(root, net_num_desks, XA_CARDINAL, 0, &ndesks, 1, NULL);
	get_atoms(root, net_cur_desk, XA_CARDINAL, 0, &cur_desk, 1, NULL);

	shape = XShapeQueryExtension(dpy, &shape_event, &shape_err);

	XQueryTree(dpy, root, &qroot, &qparent, &wins, &nwins);
	for (i = 0; i < nwins; i++) {
		XGetWindowAttributes(dpy, wins[i], &attr);
		if (!attr.override_redirect && attr.map_state == IsViewable) {
			c = new_client(wins[i]);
			c->placed = 1;
			map_client(c);
		}
	}
	XFree(wins);

	sattr.event_mask = SubMask | ColormapChangeMask | ButtonMask;
	XChangeWindowAttributes(dpy, root, CWEventMask, &sattr);
}

void
sig_handler(int signum)
{
	pid_t pid;
	int status;

	switch (signum) {
	case SIGINT:
	case SIGTERM:
	case SIGHUP:
		write(exitmsg[1], &exitmsg, 1);
		break;
	case SIGCHLD:
		while ((pid = waitpid(WAIT_ANY, &status, WNOHANG)) > 0 ||
		    (pid < 0 && errno == EINTR))
			;
		break;
	}
}

int
handle_xerror(Display * dpy, XErrorEvent * e)
{
	char msg[255];

#ifdef DEBUG
	client_t *c = find_client(e->resourceid, MATCH_WINDOW);
#endif

	if (e->error_code == BadAccess && e->resourceid == root) {
		fprintf(stderr, "%s: root window unavailable\n", __progname);
		exit(1);
	}

	XGetErrorText(dpy, e->error_code, msg, sizeof msg);
	fprintf(stderr, "%s: X error (%#lx): %s\n", __progname, e->resourceid,
	    msg);
#ifdef DEBUG
	if (c) {
#if 0 /* these can introduce additional xerrors so don't call them from here */
		dump_info(c);
		del_client(c, DEL_WITHDRAW);
#endif
	}
#endif

	return 0;
}

/* Ick. Argh. You didn't see this function. */
int
ignore_xerror(Display * dpy, XErrorEvent * e)
{
	return 0;
}

/*
 * We use XQueryTree here to preserve the window stacking order, since the
 * order in our linked list is different.
 */
static void
cleanup(void)
{
	unsigned int nwins, i;
	Window qroot, qparent, *wins;
	client_t *c;

	XQueryTree(dpy, root, &qroot, &qparent, &wins, &nwins);
	for (i = 0; i < nwins; i++) {
		c = find_client(wins[i], MATCH_FRAME);
		if (c)
			del_client(c, DEL_REMAP);
	}
	XFree(wins);

	XftFontClose(dpy, xftfont);
	XFreeCursor(dpy, map_curs);
	XFreeCursor(dpy, move_curs);
	XFreeCursor(dpy, resize_n_curs);
	XFreeCursor(dpy, resize_s_curs);
	XFreeCursor(dpy, resize_e_curs);
	XFreeCursor(dpy, resize_w_curs);
	XFreeCursor(dpy, resize_nw_curs);
	XFreeCursor(dpy, resize_sw_curs);
	XFreeCursor(dpy, resize_ne_curs);
	XFreeCursor(dpy, resize_se_curs);
	XFreeGC(dpy, pixmap_gc);
	XFreeGC(dpy, invert_gc);
	XFreePixmap(dpy, close_pm);
	XFreePixmap(dpy, close_pm_mask);
	XFreePixmap(dpy, shade_pm);
	XFreePixmap(dpy, shade_pm_mask);
	XFreePixmap(dpy, zoom_pm);
	XFreePixmap(dpy, zoom_pm_mask);
	XFreePixmap(dpy, unzoom_pm);
	XFreePixmap(dpy, unzoom_pm_mask);

	XInstallColormap(dpy, DefaultColormap(dpy, screen));
	XSetInputFocus(dpy, PointerRoot, RevertToPointerRoot, CurrentTime);

	XDeleteProperty(dpy, root, net_supported);
	XDeleteProperty(dpy, root, net_client_list);

	XCloseDisplay(dpy);
	exit(0);
}
